File: genicam/src/CameraInfo.cpp
#include <rclcpp/rclcpp.hpp>
#include <harvester_interfaces/msg/camera_device.hpp>
#include <harvester_interfaces/msg/camera_device_array.hpp>

#include <ArenaApi.h>
#include <genicam/CameraSets.hpp>

#include <genicam/CameraInfo.hpp>


uint64_t convert_mac(std::string mac) {
    mac.erase(std::remove(mac.begin(), mac.end(), ':'), mac.end());
    return strtoul(mac.c_str(), NULL, 16);
}

CameraArray::CameraArray(harvester_interfaces::msg::CameraDeviceArray::SharedPtr camera_info) : Node("camera_array"), camera_info (camera_info) {
    publish_array_info_timer_ = this->create_wall_timer(std::chrono::seconds(1), std::bind(&CameraArray::publish_camera_info, this));
    array_pub_ = this->create_publisher<harvester_interfaces::msg::CameraDeviceArray>("/caminfo", 10);
    service = this->create_service<camc>("caminfo", std::bind(&CameraArray::service_trigger, this, std::placeholders::_1, std::placeholders::_2));
}

void CameraArray::service_trigger(const std::shared_ptr<camc::Request> request, std::shared_ptr<camc::Response> response) {
    if (camera_info->cameras.empty()) {
        response->success = false;
        response->message = "No cameras found";
        return;
    }
    std::string camera_name = request->camera_name;
    std::string mac_address = request->mac_address;
    harvester_interfaces::msg::CameraDevice camera_device = harvester_interfaces::msg::CameraDevice();
    if (!mac_address.empty()) {
        uint64_t hex_mac = convert_mac(mac_address);
        for (const auto &camera : camera_info->cameras) {
            if (convert_mac(camera.mac) == hex_mac) {
                camera_device = camera;
                response->success = true;
                break;
            }
        }
    } else if (!camera_name.empty()) {
        for (const auto &camera : camera_info->cameras) {
            if (camera.name == camera_name) {
                camera_device = camera;
                response->success = true;
                break;
            }
        }
    } else {
        response->success = false;
        response->message = "Missing camera name or mac address";
    }
}

void CameraArray::publish_camera_info() {
    if (camera_info->cameras.empty()) {
        return;
    }
    array_pub_->publish(*camera_info);
    std::vector<std::string> ids;
    for (const auto &camera : camera_info->cameras) {
        ids.push_back(camera.id);
    }
}


CameraInfo::CameraInfo(harvester_interfaces::msg::CameraDeviceArray::SharedPtr camera_info) : Node("camera_info"), camera_info (camera_info) {
    update_camera_info_timer_ = this->create_wall_timer(std::chrono::minutes(1), std::bind(&CameraInfo::update_camera_info, this));

    this->declare_parameter("param_int", 0);
    this->declare_parameter("param_float", 0.0);
    this->declare_parameter("param_bool", true);
    this->declare_parameter("param_str", "default");

    param_subscriber_ = std::make_shared<rclcpp::ParameterEventHandler>(this);

    param_int = param_subscriber_->add_parameter_callback("param_int", std::bind(&CameraInfo::param_callback, this, std::placeholders::_1));
    param_str = param_subscriber_->add_parameter_callback("param_float", std::bind(&CameraInfo::param_callback, this, std::placeholders::_1));
    param_float = param_subscriber_->add_parameter_callback("param_bool", std::bind(&CameraInfo::param_callback, this, std::placeholders::_1));
    param_bool = param_subscriber_->add_parameter_callback("param_str", std::bind(&CameraInfo::param_callback, this, std::placeholders::_1));

    update_camera_info();
}

void CameraInfo::param_callback(const rclcpp::Parameter & p) {
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Param type: %s", p.get_type_name().c_str());
    if (p.get_type_name() == "string") {
        harvester_interfaces::msg::CameraDevice camera_device = harvester_interfaces::msg::CameraDevice();
        camera_device.id = p.get_parameter_value().get<std::string>();
        camera_info->cameras.push_back(camera_device);
        auto val = p.get_parameter_value().get<std::string>();
    } else if (p.get_type_name() == "integer") {
        auto val = p.get_parameter_value().get<int>();
    } else if (p.get_type_name() == "double") {
        auto val = p.get_parameter_value().get<float>();
    } else if (p.get_type_name() == "bool") {
        auto val = p.get_parameter_value().get<bool>();
    }
}


void CameraInfo::update_camera_info() {
    Arena::ISystem* pSystem = nullptr;
    std::vector<std::pair<Arena::IDevice*, uint64_t>> vDevices = std::vector<std::pair<Arena::IDevice*, uint64_t>>();
    harvester_interfaces::msg::CameraDeviceArray camera_array = harvester_interfaces::msg::CameraDeviceArray();

    try {
        pSystem = Arena::OpenSystem();
        pSystem->UpdateDevices(1000);
        std::vector<Arena::DeviceInfo> deviceInfos = pSystem->GetDevices();
        for (auto& deviceInfo : deviceInfos){
            harvester_interfaces::msg::CameraDevice camera_device = harvester_interfaces::msg::CameraDevice();
            camera_device.model = deviceInfo.ModelName();
            camera_device.vendor = deviceInfo.VendorName();
            camera_device.serial = deviceInfo.SerialNumber();
            camera_device.ip = deviceInfo.IpAddressStr();
            camera_device.subnetmask = deviceInfo.SubnetMaskStr();
            camera_device.defaultgateway = deviceInfo.DefaultGatewayStr();
            camera_device.mac = deviceInfo.MacAddressStr();
            camera_device.name = deviceInfo.UserDefinedName();
            camera_device.dhcp = deviceInfo.IsDHCPConfigurationEnabled();
            camera_device.presistentip = deviceInfo.IsPersistentIpConfigurationEnabled();
            camera_device.lla = deviceInfo.IsLLAConfigurationEnabled();

            uint64_t hex_mac = convert_mac(deviceInfo.MacAddressStr().c_str());
            if (camset::by_mac.find(hex_mac) != camset::by_mac.end()) {
                camera_device.light_ip = camset::by_mac.at(hex_mac).ip;
                camera_device.light_port = "6512";
                camera_device.light_channel = camset::by_mac.at(hex_mac).channel;
                camera_device.id = camset::by_mac.at(hex_mac).name;
            }
            camera_array.cameras.push_back(camera_device);   
        }
        Arena::CloseSystem(pSystem);
        *camera_info = camera_array;

    } catch(const std::exception& e) {
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Error: %s", e.what());
    } catch (GenICam::GenericException& ge) {
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Error: %s", ge.what());
    }
}


int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    std::cout << "... GETTING CAMERAS ..." << std::endl;

    rclcpp::executors::MultiThreadedExecutor executor;
    harvester_interfaces::msg::CameraDeviceArray::SharedPtr camera_info = std::make_shared<harvester_interfaces::msg::CameraDeviceArray>();

    auto camera_info_node = std::make_shared<CameraInfo>(camera_info);
    auto camera_array_node = std::make_shared<CameraArray>(camera_info);

    executor.add_node(camera_info_node);
    executor.add_node(camera_array_node);

    try {
        executor.spin();
    } catch (const std::exception &e) {
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "An error occurred: %s", e.what());
    }

    executor.remove_node(camera_info_node);
    executor.remove_node(camera_array_node);

    rclcpp::shutdown();
    return 0;
}

File: harvester/harvester/camera_info.py
from arena_api.system import system
from arena_api.buffer import *
from genicam.msg import CameraDevice, CameraDeviceArray
from rclpy.executors import MultiThreadedExecutor


import rclpy
from rclpy.node import Node

IP_LIGHT = {
      0 : None,
      1 : "192.168.2.251", # 7 RIGHT     70:B3:D5:DD:90:FD
      3 : "192.168.2.252", # - CENTER    70:B3:D5:DD:90:EF
      2 : "192.168.2.253", # 9 LEFT      70:B3:D5:DD:90:ED
}

BY_MAC = { 
    ##  MAC             IP      CHANNEL NAME       
    30853686642969: [IP_LIGHT[2],   7,   "9b"],
    30853686643065: [IP_LIGHT[2],   1,   "9a"],
    30853686646563: [IP_LIGHT[2],   2,   "9c"],
    30853686653149: [IP_LIGHT[2],   5,   "9d"],
    30853686643056: [IP_LIGHT[2],   6,   "9e"],
    30853686646554: [IP_LIGHT[1],   6,   "7a"],
    30853686652294: [IP_LIGHT[1],   1,   "7b"],
    30853686445113: [IP_LIGHT[1],   5,   "7c"],
    30853686646528: [IP_LIGHT[1],   2,   "7d"],
    30853686653140: [IP_LIGHT[1],   3,   "7e"],
    30853686643187: [IP_LIGHT[0],   0,   "1"],
    30853686650340: [IP_LIGHT[3],   6,   "4"],
    30853686646397: [IP_LIGHT[3],   5,   "3"],
    30853686643161: [IP_LIGHT[3],   2,   "13"],
    30853686643152: [IP_LIGHT[3],   1,   "12"],
    30853686646406: [IP_LIGHT[3],   3,   "10"],
    30853686650497: [IP_LIGHT[3],   7,   "11a"],
    30853686650366: [IP_LIGHT[3],   7,   "11b"],
}

camera_info = None

class CameraArray(Node):
    def __init__(self):
        super().__init__('camera_array')
        self.publish_array_info_pub = self.create_timer(0.001, self.publish_array_info)
        self.array_pub = self.create_publisher(CameraDeviceArray, '/caminfo', 10)

    def publish_array_info(self):
        global camera_info
        if camera_info is not None:
            self.array_pub.publish(camera_info)
            ids = [camera.id for camera in camera_info.cameras]
            # self.get_logger().info(f"cameras infos of {ids} are published")

class CameraInfo(Node):
    def __init__(self):
        super().__init__('camera_info')
        self.update_camera_info_pub = self.create_timer(30, self.update_camera_info)
        self.update_camera_info()

    """
    A list of dictionaries used to create devices. Each dictionary represents a discovered device on the network.
    A device info dictionary has the following keys: `model`,`vendor`, `serial`, `ip`, `subnetmask`,
    `defaultgateway`, `mac`, `name`, `dhcp`, `presistentip`, `lla`, and `version`. 
    """
    
    def update_camera_info(self):
        global camera_info
        device_infos_no_dup = []
        for device in system.device_infos:
            if device not in device_infos_no_dup:
                device_infos_no_dup.append(device)
        camera_array = CameraDeviceArray()
        for device in device_infos_no_dup:
            camera_device = CameraDevice()
            camera_device.model = device["model"]
            camera_device.vendor = device["vendor"]
            camera_device.serial = device["serial"]
            camera_device.ip = device["ip"]
            camera_device.subnetmask = device["subnetmask"]
            camera_device.defaultgateway = device["defaultgateway"]
            camera_device.mac = device["mac"]
            camera_device.name = device["name"]
            camera_device.dhcp = device["dhcp"]
            camera_device.presistentip = device["presistentip"]
            camera_device.lla = device["lla"]
            camera_device.version = device["version"]

            hex_mac = int(device["mac"].replace(":", ""), 16)
            if hex_mac in BY_MAC:
                camera_device.light_ip = BY_MAC[hex_mac][0]
                camera_device.light_port = "6512"
                camera_device.light_channel = str(BY_MAC[hex_mac][1])
                camera_device.id = BY_MAC[hex_mac][2]

            camera_array.cameras.append(camera_device)
        camera_info = camera_array
        self.get_logger().info(f"camera info is updated, {len(camera_array.cameras)} cameras are found")
        

def main(args=None):
    rclpy.init()
    print('... GETTING CAMERAS ...')
    executor = MultiThreadedExecutor(num_threads=2)
    camera_info_node = CameraInfo()
    camera_array_node = CameraArray()
    try:
        executor.add_node(camera_info_node)
        executor.add_node(camera_array_node)
        try:
            executor.spin()
        except Exception as e:
            print(f"An error occurred: {str(e)}")
    except Exception as e:
        print(f"An error occurred: {str(e)}")
    finally:
        executor.remove_node(camera_info_node)
        executor.remove_node(camera_array_node)
        executor.shutdown()
        rclpy.shutdown()


if __name__ == '__main__':
    main()